source/extensions/watchdog/profile_action/profile_action.cc

```diff
@@ -0,0 +1,115 @@
+    : path_(config.profile_path()), running_profile_(false),
+      max_profiles_per_tid_(config.max_profiles_per_thread() == 0
+                                ? DefaultMaxProfilePerTid
+                                : config.max_profiles_per_thread()),
+      profiles_started_(0), duration_(std::chrono::milliseconds(
+                                PROTOBUF_GET_MS_OR_DEFAULT(config, profile_duration, 5000))),
+      context_(context), timer_cb_(nullptr) {}
+
+void ProfileAction::run(
+    envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent /*event*/,
+    const std::vector<std::pair<Thread::ThreadId, MonotonicTime>>& thread_ltt_pairs,
+    MonotonicTime /*now*/) {
+  if (running_profile_) {
+    return;
+  }
+
+  // Check if there's a tid that justifies profiling
+  auto trigger_tid = getTidTriggeringProfile(thread_ltt_pairs);
+  if (!trigger_tid.has_value()) {
+    ENVOY_LOG_MISC(warn, "None of the provide tids justify profiling");
+    return;
+  }
+
+  auto& fs = context_.api_.fileSystem();
+  if (!fs.directoryExists(path_)) {
+    ENVOY_LOG_MISC(error, "Directory path {} doesn't exist.", path_);
+    return;
+  }
+
+  // Generate file path for output and try to profile
+  profile_filename_ = generateProfileFilePath(path_, context_.api_.timeSource().systemTime());
+
+  if (!Profiler::Cpu::profilerEnabled()) {
+    if (Profiler::Cpu::startProfiler(profile_filename_)) {
+      // Update state
+      running_profile_ = true;
+      ++profiles_started_;
+      tid_to_profile_count_[*trigger_tid] += 1;
+
+      // Schedule callback to stop
+      if (timer_cb_ == nullptr) {
+        // Create the timer once.
+        timer_cb_ = context_.dispatcher_.createTimer([this] {
+          if (Profiler::Cpu::profilerEnabled()) {
+            Profiler::Cpu::stopProfiler();
+            running_profile_ = false;
+          } else {
+            ENVOY_LOG_MISC(error,
+                           "Profile Action's stop() was scheduled, but profiler isn't running!");
+          }
+
+          if (!context_.api_.fileSystem().fileExists(profile_filename_)) {
+            ENVOY_LOG_MISC(error, "Profile file {} wasn't created!", profile_filename_);
+          }
+        });
+      }
+
+      timer_cb_->enableTimer(duration_);
+    } else {
+      ENVOY_LOG_MISC(error, "Profile Action failed to start the profiler.");
+    }
+  } else {
+    ENVOY_LOG_MISC(error, "Profile Action unable to start the profiler as it is in use elsewhere.");
+  }
+}
+
+// Helper to determine if we have a valid tid to start profiling.
+absl::optional<Thread::ThreadId> ProfileAction::getTidTriggeringProfile(
+    const std::vector<std::pair<Thread::ThreadId, MonotonicTime>>& thread_ltt_pairs) {
+  absl::optional<Thread::ThreadId> tid_to_profile;
+
+  // Find a TID not over the max_profiles threshold
+  for (const auto& [tid, ltt] : thread_ltt_pairs) {
+    if (tid_to_profile_count_[tid] < max_profiles_per_tid_) {
+      tid_to_profile = tid;
+      break;
```

Comment:
Can't this just be `return tid;` and `return absl::nullopt;` at the bottom? That lets you get rid of the extra variable.

Reply:
Done.